src/node/rpc/frontend.h

```diff
@@ -436,6 +436,13 @@ namespace ccf
             }
           }
         }
+        catch (const kv::CompactedVersionConflict& e)
+        {
+          // The executing transaction failed because of a conflicting
+          // compaction. Reset and retry
+          tx.reset();
```

Comment:
I might be wrong but this concerns me for a bit.
If a tx that fails with CompactedVersionConflict takes a minute to run and we compact every 10 seconds then we could end up in a scenario where this tx will realistically never commit but eat system resources (Assuming this is some background load in the system). Should we consider failing this transaction after X retries?

Reply:
Wrote a long blurb here but I changed my mind. In short: This is similar to the existing conflict case (`kv::CommitSuccess::CONFLICT`). We only get conflicts in either case if another transaction write transaction is submitted while we're executing (if there are no writes since the last compact, this map is unaffected by the compact, so the requested 'old' version is still available). So any transaction that takes 1 minute _may_ never complete, but only if other transactions are submitted in the interim.

I think a retry limit is a sensible fix for either case. Perhaps this should be caught by a timeout instead - if a transaction takes too long (because it was forwarded, or because it was retried 10 times, or just because its really slow) we may want some global system to kill it and return an error to the user? Let's discuss at standup.

Reply:
+1 for a retry cap.

Also, an explicit mention in the doc to tell users _not_ to catch all/or `CompactedVersionConflict` in code that calls get_view()...

Reply:
I'll add a retry cap.

With regard to avoiding accidental catches, I've made `CompactedVersionConflict` not inherit from any standard exception type, so its slightly harder to catch accidentally (you'd have to do the unusual `catch (...)` rather than `catch (const std::exception& e)`). I wonder if we should just avoid mentioning this particular foot-cannon in the docs at all, rather than saying "please don't load it". To expand on that: I'm not sure where this would live in the doc, and it opens the door to a wider class of hard-to-enumerate-but-incredibly-dangerous things people _could_ do in the C++ API (eg - creating their own `Tx`, calling `commit()`, flipping bits in the kv). In particular, they could already break opacity by talking about mid-execution transaction results in member fields of the app frontend, and I think that catching this (and ignoring it!?) would only allow opacity breaches and nothing else?

Reply:
I briefly tried searching for `catch (...)` but quickly gave up for obvious and yet sad reasons. I'm not sure it's as rare as that (it was certainly plentiful in some code bases I've worked on that used a mix of RogueWave and STL), but I can't prove it. As often/always, it's not so much someone writing this that's likely to happen first, it's someone using a library that does, and passing a callback/functor, which will be unexpectedly wrapped in `catch (...)`. That's why it's often worth having an FAQ keyed by symptoms and offering helpful suggestions. "You see weird thing A:  check you've got the right B".

It may be good to at least spell out positive guidelines/best practices such as keep all state in the KV and nowhere else, I'm not sure that avoiding talking about the limitations of our programming model altogether is such a good thing. It also seems to me that mentioning the existence of the foot gun in an open source project creates more opportunity for a passer-by to propose a way to remove it.

Anyway, that's all fluffy opinion, I've got no facts. I do think making not inherit a standard exception type is an excellent idea in any case.